### ISSUE TITLE
Liquibase Core Version 4.18 taking long time in computing checksum

### ISSUE THREAD

**Issue by Sumitkrchaudhary** [liquibase/liquibase#3694]
## Environment

**Liquibase Version**:
4.18
**Liquibase Integration & Version**: <Pick one: CLI, maven, gradle, spring boot, servlet, etc.>
CLI
**Liquibase Extension(s) & Version**: 
4.18
**Database Vendor & Version**:
Oracle Database 19c Enterprise Edition Release  19.3.0.0.0                                                     

**Operating System Type & Version**:
Oracle Enterprise Linux 7
**Infrastructure Type/Provider**: <AWC, GCS, Azure, VM, etc>
VM

## Description
Starting Liquibase at 20:06:21 (version 4.17.2 #5255 built at 2022-11-01 18:07+0000)
[2022-11-30 20:06:23] INFO [liquibase.lockservice] Successfully acquired change log lock
[2022-11-30 20:06:39] INFO [liquibase.changelog] Creating database history table with name: TEST_APP.DBCHANGELOG
[2022-11-30 20:06:39] INFO [liquibase.changelog] Reading from TEST_APP.DBCHANGELOG
Running Changeset: changelogs/prerequisite_checks.xml::1::base
[2022-11-30 20:29:36] INFO [liquibase.changelog] ChangeSet changelogs/prerequisite_checks.xml::1::base ran successfully in 16ms

Starting Liquibase at 23:59:58 (version 4.18.0 #5864 built at 2022-12-02 18:02+0000)
[2023-01-09 23:59:59] INFO [liquibase.lockservice] Successfully acquired change log lock
[2023-01-10 00:00:18] INFO [liquibase.changelog] Creating database history table with name: TEST_APP.DBCHANGELOG
[2023-01-10 00:00:18] INFO [liquibase.changelog] Reading from TEST_APP.DBCHANGELOG
Running Changeset: changelogs/prerequisite_checks.xml::1::base
[2023-01-10 00:22:13] INFO [liquibase.changelog] ChangeSet changelogs/prerequisite_checks.xml::1::base ran successfully in 16ms

Taking approx ~ 22 mins in version 4.17.2 and version 4.18.0

Starting Liquibase at 16:24:35 (version 4.14.0 #3667 built at 2022-07-22 18:36+0000)
[2022-07-29 16:24:37] INFO [liquibase.lockservice] Successfully acquired change log lock
[2022-07-29 16:24:51] INFO [liquibase.changelog] Creating database history table with name: TEST_APP.DBCHANGELOG
[2022-07-29 16:24:51] INFO [liquibase.changelog] Reading from TEST_APP.DBCHANGELOG
Running Changeset: changelogs/prerequisite_checks.xml::1::base
[2022-07-29 16:25:04] INFO [liquibase.changelog] ChangeSet changelogs/prerequisite_checks.xml::1::base ran successfully in 16ms

Taking approx ~ few secs in version 4.14.0  ( issue not seen here)

A clear and concise description of the issue being addressed.
- Describe the actual problematic behavior.
- Ensure private information is redacted.

## Steps To Reproduce
1. Run Liquibase version 4.14, runs fine no slowness seen
2. Run Liquibase version 4.17.2 and version 4.18, takes long time as shown above, Issue seen in recent versions.
 
List the steps to reproduce the behavior.
1. Run Liquibase version 4.14, runs fine no slowness seen
2. Run Liquibase version 4.17.2 and version 4.18, takes long time as shown above, Issue seen in recent versions.

## Actual Behavior
1. Run Liquibase version 4.14, runs fine no slowness seen
2. Run Liquibase version 4.17.2 and version 4.18, takes long time as shown above, Issue seen in recent versions.

## Expected/Desired Behavior
1. Run Liquibase version 4.17.2 and version 4.18, no slowness should be seen as mentioned in above steps. 
This should behave as Liquibase version 4.14.

## Additional Context

Jars used under /lib folder in version 4.14, version 4.17.2 and version 4.18
commons-lang3.jar
liquibase-core.jar
ojdbc8.jar
opencsv.jar
README.txt
snakeyaml.jar

Version 4.18 and 4.17.2 taking long time in computing checksum

**Comment by kevin-atx** [comment]
Hey @filipelautert - could you review this issue and see if it's a duplicate of https://github.com/liquibase/liquibase/issues/3675 ?

**Comment by filipelautert** [comment]
@kevin-atx seems so.

Hello @Sumitkrchaudhary - thanks for the report!
We are reviewing this issue and I think we may find the point were there the problem is. I implemented a simple test fix on PR https://github.com/liquibase/liquibase/pull/3684 . Here are the artifacts for it: https://github.com/liquibase/liquibase/suites/10339006499/artifacts/509167433

Could you please use those artifacts and check if the problem is resolved ?

Thanks

**Comment by Sumitkrchaudhary** [comment]
Taking uploaded artifacts and will confirm.

**Comment by Sumitkrchaudhary** [comment]
The same problem is with Version 4.19 as well

**Comment by Sumitkrchaudhary** [comment]
This seems working fast and well with new artifacts uploaded above. Looking at faster execution time seems your fix is working well. :)

Here new execution time which shows faster execution -

For new database setup with 4.18 new uploaded artifacts --
Starting Liquibase at 17:37:02 (version [Core: //3675-performance-regression-on-update-in-417/6576/c0a86e/2023-01-12 18:24+0000] #6576 built at 2023-01-12 18:24+0000)
[2023-02-06 17:37:04] INFO [liquibase.lockservice] Successfully acquired change log lock
[2023-02-06 17:37:14] INFO [liquibase.changelog] Creating database history table with name: TEST_APP.DBCHANGELOG
[2023-02-06 17:37:14] INFO [liquibase.changelog] Reading from TEST_APP.DBCHANGELOG
Running Changeset: changelogs/prerequisite_checks.xml::1::base
[2023-02-06 17:37:14] INFO [liquibase.changelog] ChangeSet changelogs/prerequisite_checks.xml::1::base ran successfully in 37ms 

Doing some more regressions and will update.

**Comment by Sumitkrchaudhary** [comment]
This behaviour of slowness in upgrading existing db is similar to Version 4.14 as mentioned in issue#
https://github.com/liquibase/liquibase/issues/3655

**Comment by Sumitkrchaudhary** [comment]
Is there any artifact that can be shared for this fix that I can use to confirm

**Comment by Sumitkrchaudhary** [comment]
artifact on version 4.19 works better if that can be shared

**Comment by filipelautert** [comment]
Hello @Sumitkrchaudhary! Thanks for the help. You can use the artifacts from the following link, they are the latest ones built from master: https://github.com/liquibase/liquibase/suites/11087994301/artifacts/564759242

**Comment by Sumitkrchaudhary** [comment]
===================================================
Starting Liquibase at 14:36:01 (version [Core: //master/7474/0b2025/2023-02-20 17:24+0000] #7474 built at 2023-02-20 17:24+0000)
**---first** 
[2023-03-06 14:36:22] INFO [liquibase.changelog] Reading from TEST_APP.DBCHANGELOG
[2023-03-06 15:18:24] INFO [liquibase.lockservice] Successfully acquired change log lock
**---second** 
[2023-03-06 15:18:26] INFO [liquibase.changelog] Reading from TEST_APP.DBCHANGELOG
[2023-03-06 15:49:51] INFO [liquibase.changelog] Reading from TEST_APP.DBCHANGELOG
[2023-03-06 15:54:03] INFO [liquibase.lockservice] Successfully released change log lock

---hub is not propmted now with liquibase.hub.mode=off
Do you want to see this operation's report in Liquibase Hub, which improves team collaboration? 
If so, enter your email. If not, enter [N] to no longer be prompted, or [S] to skip for now, but ask again next time [S]: 
[2023-03-06 15:54:03] INFO [liquibase.lockservice] Successfully acquired change log lock
Skipping auto-registration
[2023-03-06 15:54:03] WARNING [liquibase.hub] Skipping auto-registration
**---third** 
[2023-03-06 15:54:03] INFO [liquibase.changelog] Reading from TEST_APP.DBCHANGELOG
Running Changeset: changelogs/prerequisite_checks.xml::1::base
[2023-03-06 16:29:56] INFO [liquibase.changelog] ChangeSet changelogs/prerequisite_checks.xml::1::base ran successfully in 15ms

Running Changeset: changelogs/prerequisite_checks.xml::2::base
[2023-03-06 16:29:56] INFO [liquibase.changelog] ChangeSet changelogs/prerequisite_checks.xml::2::base ran successfully in 15ms

With this artifact when ran upgrading of existing db, the performance issue is still seen but a new problem introduced is it is reading DBCHANGELOG multiple times (3 times as shown in above logs) now which is making it even worse in terms of overall upgrade time.

**Comment by Sumitkrchaudhary** [comment]
[2023-03-06 21:25:47] INFO [liquibase.lockservice] Successfully acquired change log lock
[2023-03-06 21:25:47] INFO [liquibase.changelog] Creating database history table with name: TEST_APP.DBCHANGELOG
[2023-03-06 21:25:47] INFO [liquibase.changelog] Reading from TEST_APP.DBCHANGELOG
Running Changeset: changelogs/prerequisite_checks.xml::1::base
[2023-03-06 21:46:55] INFO [liquibase.changelog] ChangeSet changelogs/prerequisite_checks.xml::1::base ran successfully in 32ms

Took around ~ 21 mins on DBCHANGELOG.

With this artifact when setting up new fresh db, the performance issue is still seen while reading DBCHANGELOG. With previous artifact this was not seen.

**Comment by filipelautert** [comment]
Hello! We released 4.19.1 last week, could you check this one too?
Also, what kind of check do you have in changelogs/prerequisite_checks.xml::1::base ?  are those only preConditions tags? do you perform any explicit query to validate DBCHANGELOG table?

**Comment by Sumitkrchaudhary** [comment]
This is a check with a very basic sql - changelogs/prerequisite_checks.xml::1::base
	
	<changeSet id="1" author="base" runAlways="true" >
		<preConditions onFail="HALT" onFailMessage="Prerequisite Failed: DB instance NLS_LENGTH_SEMANTICS is not CHAR" >
			<sqlCheck expectedResult="CHAR">SELECT value FROM nls_instance_parameters WHERE parameter = 'NLS_LENGTH_SEMANTICS'</sqlCheck>
		</preConditions>
		<comment>DB session NLS_LENGTH_SEMANTICS is CHAR.</comment>
    </changeSet>

I do not perform any query to validate DBCHANGELOG table.

Everything works very efficiently in Version 3.8.8 which is used currently.

**Comment by Sumitkrchaudhary** [comment]
In case of upgrading existing database with version 4.20.0

Starting Liquibase at 03:02:51 (version 4.20.0 #7837 built at 2023-03-07 16:25+0000)
---first
[2023-03-11 03:03:09] INFO [liquibase.changelog] Reading from TEST_APP.DBCHANGELOG
[2023-03-11 03:03:10] INFO [liquibase.executor] Changelog query completed.
[2023-03-11 04:12:31] INFO [liquibase.executor] Changelog query completed.
[2023-03-11 04:12:31] INFO [liquibase.executor] Changelog query completed.
[2023-03-11 04:12:31] INFO [liquibase.executor] Changelog query completed.
[2023-03-11 04:12:31] INFO [liquibase.lockservice] Successfully acquired change log lock
[2023-03-11 04:12:33] INFO [liquibase.executor] Changelog query completed.

---second
[2023-03-11 04:12:33] INFO [liquibase.changelog] Reading from TEST_APP.DBCHANGELOG
[2023-03-11 04:12:34] INFO [liquibase.executor] Changelog query completed.
[2023-03-11 04:45:52] INFO [liquibase.executor] Changelog query completed.

---third
[2023-03-11 04:45:52] INFO [liquibase.changelog] Reading from TEST_APP.DBCHANGELOG
[2023-03-11 04:45:53] INFO [liquibase.executor] Changelog query completed.
Running Changeset: changelogs/prerequisite_checks.xml::1::base
[2023-03-11 05:27:36] INFO [liquibase.changelog] ChangeSet changelogs/prerequisite_checks.xml::1::base ran successfully in 31ms

With this version 4.20.0 when ran upgrading of existing db, the performance issue is still seen but a new problem introduced as seen in artifact it is reading DBCHANGELOG multiple times (3 times as shown in above logs) now which is making it even worse in terms of overall upgrade time.




Starting Liquibase at 02:22:04 (version 4.20.0 #7837 built at 2023-03-07 16:25+0000)
[2023-03-11 02:22:26] INFO [liquibase.executor] Changelog query completed.
[2023-03-11 02:22:26] INFO [liquibase.executor] Changelog query completed.
[2023-03-11 02:22:26] INFO [liquibase.executor] Changelog query completed.
[2023-03-11 02:22:26] INFO [liquibase.executor] Changelog query completed.
[2023-03-11 02:22:26] INFO [liquibase.executor] Changelog query completed.
[2023-03-11 02:22:26] INFO [liquibase.lockservice] Successfully acquired change log lock
[2023-03-11 02:22:27] INFO [liquibase.changelog] Creating database history table with name: TEST_APP.DBCHANGELOG
[2023-03-11 02:22:27] INFO [liquibase.executor] Changelog query completed.
[2023-03-11 02:22:27] INFO [liquibase.changelog] Reading from TEST_APP.DBCHANGELOG
[2023-03-11 02:22:27] INFO [liquibase.executor] Changelog query completed.
[2023-03-11 02:22:27] INFO [liquibase.executor] Changelog query completed.
Running Changeset: changelogs/prerequisite_checks.xml::1::base
[2023-03-11 02:44:31] INFO [liquibase.changelog] ChangeSet changelogs/prerequisite_checks.xml::1::base ran successfully in 31ms
[2023-03-11 02:44:31] INFO [liquibase.executor] Changelog query completed.
[2023-03-11 02:44:31] INFO [liquibase.executor] Changelog query completed.
Running Changeset: changelogs/prerequisite_checks.xml::2::base
[2023-03-11 02:44:31] INFO [liquibase.changelog] ChangeSet changelogs/prerequisite_checks.xml::2::base ran successfully in 16ms
[2023-03-11 02:44:31] INFO [liquibase.executor] Changelog query completed.


Took around ~ 22 mins on DBCHANGELOG.

With this version 4.20 as seen in artifact also when setting up new fresh db, the performance issue is still seen while reading DBCHANGELOG. With previous first artifact provided in the starting this was not seen.

**Comment by Sumitkrchaudhary** [comment]
In case you need a demo of this issue I can show and explain. Let me know if it is possible

**Comment by Sumitkrchaudhary** [comment]
The same proble is faced in version 4.19.1 as shown in below logs:

Starting Liquibase at 03:15:16 (version 4.19.1 #7641 built at 2023-02-28 18:08+0000)
[2023-03-11 03:15:52] INFO [liquibase.executor] Changelog query completed.
[2023-03-11 03:15:52] INFO [liquibase.executor] Changelog query completed.
[2023-03-11 03:15:52] INFO [liquibase.executor] Changelog query completed.
[2023-03-11 03:15:52] INFO [liquibase.executor] Changelog query completed.
[2023-03-11 03:15:52] INFO [liquibase.executor] Changelog query completed.
[2023-03-11 03:15:52] INFO [liquibase.lockservice] Successfully acquired change log lock
[2023-03-11 03:15:53] INFO [liquibase.changelog] Creating database history table with name: TEST_APP.DBCHANGELOG
[2023-03-11 03:15:53] INFO [liquibase.executor] Changelog query completed.
[2023-03-11 03:15:53] INFO [liquibase.changelog] Reading from TEST_APP.DBCHANGELOG
[2023-03-11 03:15:53] INFO [liquibase.executor] Changelog query completed.
[2023-03-11 03:15:53] INFO [liquibase.executor] Changelog query completed.
Running Changeset: changelogs/prerequisite_checks.xml::1::base
[2023-03-11 03:48:01] INFO [liquibase.changelog] ChangeSet changelogs/prerequisite_checks.xml::1::base ran successfully in 31ms
[2023-03-11 03:48:01] INFO [liquibase.executor] Changelog query completed.
[2023-03-11 03:48:01] INFO [liquibase.executor] Changelog query completed.
Running Changeset: changelogs/prerequisite_checks.xml::2::base
[2023-03-11 03:48:01] INFO [liquibase.changelog] ChangeSet changelogs/prerequisite_checks.xml::2::base ran successfully in 13ms
[2023-03-11 03:48:01] INFO [liquibase.executor] Changelog query completed.

Took around ~ 32 mins on DBCHANGELOG.

**Comment by filipelautert** [comment]
Hello @Sumitkrchaudhary - do you mind running it in debug (--log-level=debug) and send me the results at flautert@liquibase.org ? Then we can take from there.

**Comment by dm-ion** [comment]
Adding another data point here, for us:
On 4.16.1 our liquibase execution time was ~1 minute
On 4.20.0  our liquibase execution time was ~23 minutes

No other changes, wasn't able to test on any versions in between due to various Liquibase bugs that effected us.

Didn't investigate at all, but I did quickly pop it open in the profiler and saw in 4.20.0 an enormous amount of time is taken up by DatabaseChangeLog.normalizePath, in 4.16.1 this is not the case.

**Comment by filipelautert** [comment]
Thanks @dm-ion ! We have pr https://github.com/liquibase/liquibase/pull/3853 that aims to improve performance on normalizePath. We still testing and reviewing it, but hopefully by end of the week it will make way to master. 
Meanwhile would you be able to use the jar from it (it is available here -> https://github.com/liquibase/liquibase/actions/runs/4421668736 ) and see if the performance issue is gone or if it shifts to another place?

Thanks!

**Comment by Sumitkrchaudhary** [comment]
> Hello @Sumitkrchaudhary - do you mind running it in debug (--log-level=debug) and send me the results at [flautert@liquibase.org](mailto:flautert@liquibase.org) ? Then we can take from there.

yes, will send debug logs

**Comment by dm-ion** [comment]
> Thanks @dm-ion ! We have pr #3853 that aims to improve performance on normalizePath. We still testing and reviewing it, but hopefully by end of the week it will make way to master. Meanwhile would you be able to use the jar from it (it is available here -> https://github.com/liquibase/liquibase/actions/runs/4421668736 ) and see if the performance issue is gone or if it shifts to another place?
> 
> Thanks!

Hi, based on results from my development environment normalize path seems much cheaper on the branch. I won't be able to test it on the dataset I referenced above until it's in a released version, but it seems promising.

**Comment by filipelautert** [comment]
Thanks @dm-ion ! The pr was merged to master and it will be available when we release the next version.

**Comment by dm-ion** [comment]
I'll report back timing on that dataset once it's available. Thanks!

**Comment by Sumitkrchaudhary** [comment]
> 

Filipe: Sent to you the debug logs

**Comment by Sumitkrchaudhary** [comment]
> > Hello @Sumitkrchaudhary - do you mind running it in debug (--log-level=debug) and send me the results at [flautert@liquibase.org](mailto:flautert@liquibase.org) ? Then we can take from there.
> 
> yes, will send debug logs

Filipe: Sent to you the debug logs

**Comment by dm-ion** [comment]
Hi we were able to test this new release on that same dataset and the performance issues were completely resolved for us.

Unfortunately though we haven't been able to upgrade to the latest release yet due to another liquibase issue we haven't had a chance to investigate yet. It seems the locks aren't being released properly after the update command due to the underlying JDBC connection being closed. We'll open a ticket for this if necessary; however, it could be something with our integration.

Thanks!

**Comment by filipelautert** [comment]
Thanks @dm-ion that's good to hear! We found a problem related to lockService -> https://github.com/liquibase/liquibase/pull/4151 .  Do you think you could make a try with the fix from https://github.com/liquibase/liquibase/actions/runs/4747384342?pr=4151 ?

**Comment by filipelautert** [comment]
@Sumitkrchaudhary could you help to test with the fix I just sent in the comment above? seems we are getting there.

**Comment by Sumitkrchaudhary** [comment]
@filipelautert I see Status - Failure. Can you please confirm

**Comment by Sumitkrchaudhary** [comment]
@filipelautert any further update on this

**Comment by Sumitkrchaudhary** [comment]
@filipelautert hi, were you able to resume on this analysis further

**Comment by filipelautert** [comment]
Hello @Sumitkrchaudhary, I may get back to it tomorrow or Friday - we just merged some big stuff to master and now I've some time to check this again.

**Comment by Sumitkrchaudhary** [comment]
@filipelautert Any further update on this

**Comment by Sumitkrchaudhary** [comment]
@filipelautert waiting on any update on this as this is a blocker for us to move to latest version.

**Comment by Sumitkrchaudhary** [comment]
@filipelautert any update on this please

**Comment by Sumitkrchaudhary** [comment]
I tried with 4.19.1 and 4.20 also but the problem persists. Already published data above in this issue.

**Comment by Sumitkrchaudhary** [comment]
@kevin-atx let me know if any information is needed. Waiting on further update on this

**Comment by Sumitkrchaudhary** [comment]
@kevin-atx @filipelautert Waiting on further update on this

**Comment by Sumitkrchaudhary** [comment]
@kevin-atx @filipelautert we have to upgrade as soon possible when this issue is fixed. waiting on any more details on this.

**Comment by Sumitkrchaudhary** [comment]
> Hello @Sumitkrchaudhary - we are doing some performance improvement work on PR #4798 - this one looks promissing. It will be included on Liquibase 4.24.0

Hello, @filipelautert is their any artifact that I can use to verify as 4.24 is yet not released

**Comment by lalo-mx** [comment]
> > Hello @Sumitkrchaudhary - we are doing some performance improvement work on PR #4798 - this one looks promissing. It will be included on Liquibase 4.24.0
> 
> Hello, @filipelautert is their any artifact that I can use to verify as 4.24 is yet not released

@Sumitkrchaudhary 4.24.0 is released

**Comment by Sumitkrchaudhary** [comment]
@filipelautert This looks really good with 4.25 release. Seems working now as normal and fast.

**Comment by filipelautert** [comment]
I'm closing it. Please reopen if it happens again.
